### PR TITLE
Replication throttling

### PR DESCRIPTION
**Pull request by martinfurmanski**

### CODE REVIEW

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/ReplicatedContent.java [hunk 1]
```diff
@@ -24,4 +24,13 @@
...
 {
+    default boolean hasSize()
+    {
+        return false;
+    }
+
+    default long size()
+    {
+        throw new UnsupportedOperationException();
+    }
```
Comment (andrewkerr9000) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/ReplicatedContent.java:35]: Not sure about these names. Maybe `hasPayload` and `payloadSize`?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/SenderService.java [hunk 2]
```diff
@@ -78,12 +75,17 @@ public void send( AdvertisedSocketAddress to, Message message )
...
         }
         finally
         {
             serviceLock.readLock().unlock();
         }
+
+        if ( block )
+        {
+            future.awaitUninterruptibly();
+        }
```
Comment (andrewkerr9000) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/SenderService.java:88]: I'm not familiar with this (or Netty `Future`'s) code so I'm missing context, but I'd normally expect an async service to return a `Future`, allowing the calling site to decide whether to chain callbacks, block, or just ignore.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/Throttler.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.causalclustering.core.replication;
+
+import org.neo4j.function.ThrowingSupplier;
+
+/**
+ * Throttles calls to underlying invocation based on available credits.
+ */
+public class Throttler
```
Comment (andrewkerr9000) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/Throttler.java]: Worth unit testing?

**Review by andrewkerr9000 — COMMENTED** [review]
I can't be of much help here I'm afraid.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/SenderService.java (hunk 2 above)
Comment (martinfurmanski) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/SenderService.java:88]: Good observation and I actually was going for that initially but decided to just contain it with a simple block flag. I think there is a lot of potential for future maturity to be gained in the messaging stack. I am trying to be minimally invasive here.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/Throttler.java (hunk 3 above)
Comment (martinfurmanski) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/Throttler.java]: Definitely should be. It is on the TODO.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/ReplicatedContent.java (hunk 1 above)
Comment (martinfurmanski) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/ReplicatedContent.java:35]: Yea the naming can probably be improved. I guess everything has a payload though.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/SenderService.java (hunk 2 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/SenderService.java:88]: :+1:

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/Throttler.java (hunk 3 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/replication/Throttler.java]: :+1:

**Pull request merged**